RANDR extension upgrade to 1.5

sunweaver opened the pull request:
This PR backports latest RANDR code from Xorg's Xserver to nx-X11/programs/Xserver.

sunweaver commented:
@uli42: Here comes the promised RANDR 1.2 -> 1.5 upgrade.

uli42 commented:
On Fri, Jun 10, 2016 at 12:18 PM, Mike Gabriel notifications@github.com wrote:

> Copy x-indent.sh and x-indent-all.sh from xorg/util/modular.
> randr extension: Convert to Xorg coding style.

While I appreciate that it also brings some problems in respect to
later nx 3.7 integration. We should restrict coding style changes and
reformatting to backported code like randr1.5 and NOT use that on
other code. Otherwise it will require much more manual work when
upgrading nx/xorg7.0 to 7.1 to 7.2...

Uli

sunweaver commented:
Hi Uli,

On  Fr 10 Jun 2016 13:12:53 CEST, Ulrich Sibiller wrote:

> On Fri, Jun 10, 2016 at 12:18 PM, Mike Gabriel  
> notifications@github.com wrote:
> 
> > Copy x-indent.sh and x-indent-all.sh from xorg/util/modular.
> > randr extension: Convert to Xorg coding style.
> 
> While I appreciate that it also brings some problems in respect to
> later nx 3.7 integration. We should restrict coding style changes and
> reformatting to backported code like randr1.5 and NOT use that on
> other code. Otherwise it will require much more manual work when
> upgrading nx/xorg7.0 to 7.1 to 7.2...
> 
> Uli

I full agree here. The idea is...
- no code style changes for hw/nxagent/*
- before backporting Xorg code to a specific part
   of the Xserver: run x-indent.sh on those files
   before the backport, then backport
- no code style changes on non-backported parts
   of the Xserver

As you can see with this RANDR backport, you can easily

(a) look at the diff between nx-libs before the backport and after [1](b) look at an easy to read diff between the original X.org and the
     same files backported for nx-libs

Mike

[1]  

## https://github.com/ArcticaProject/nx-libs/pull/137/commits/48e418d4380a3871d462da5bca477aa635da059d

DAS-NETZWERKTEAM
mike gabriel, herweg 7, 24357 fleckeby
mobile: +49 (1520) 1976 148
landline: +49 (4354) 8390 139

GnuPG Key ID 0x25771B31
mail: mike.gabriel@das-netzwerkteam.de, http://das-netzwerkteam.de

freeBusy:
https://mail.das-netzwerkteam.de/mailxchange/kronolith/fb.php?u=m.gabriel%40das-netzwerkteam.de

sunweaver commented:
@uli42: Please consider reviewing this PR, as well. It is big and we shall consider your review work on this as paid time. If you feel, that this is too much work or not doable for you, please let me know.

sunweaver commented:
The PR branch of this PR still needs to be rebased against 3.6.x and updated. I'll let you know when this is done.

uli42 commented on nx-X11/programs/Xserver/hw/nxagent/Extensions.c:
```diff
@@ -114,13 +163,64 @@ void nxagentInitRandRExtension(ScreenPtr pScreen)
 {
-  fprintf(stderr, "RANDR going down - NX version\n");
+  /* FALSE means: not supported */
+  fprintf(stderr, "nxagentRandRReplaceScanoutPximap: NX's RANDR does not support scan-out pixmaps.\n");
```
typo in function name in fprintf().

Also, do we really need that message? It should be sufficient for the calling party to get the return code.

sunweaver replied:
Typo fixed locally. All new RandR printf messages have been embraced by #ifdef DEBUG now.

uli42 commented on nx-X11/programs/Xserver/hw/nxagent/Extensions.c:
```diff
@@ -114,13 +163,64 @@ void nxagentInitRandRExtension(ScreenPtr pScreen)
+   * FIXME: Add more modes per output than the current window size.
+   *        At least when in fullscreen mode.
+   */
+  fprintf(stderr, "nxagentRandRGetPanning: RANDR Panning is currently not supported.\n");
```
same here. Maybe encapsulate those message in #ifdef DEBUG...

uli42 commented on nx-X11/programs/Xserver/hw/nxagent/Extensions.c:
```diff
@@ -114,13 +163,64 @@ void nxagentInitRandRExtension(ScreenPtr pScreen)
+   * FIXME: Add more modes per output than the current window size.
+   *        At least when in fullscreen mode.
+   */
+  fprintf(stderr, "nxagentRandRSetPanning: RANDR Panning is currently not supported.\n");
```
and here

uli42 commented on nx-X11/programs/Xserver/hw/nxagent/NXextension.c:
```diff
@@ -296,7 +296,8 @@ CloseDownExtensions()
     {
-	(* extensions[i]->CloseDown)(extensions[i]);
+	if (extensions[i]->CloseDown)
+	    (* extensions[i]->CloseDown)(extensions[i]);
```
Hmm, same code as in dix/extension.c. Can't this be simplified somehow?

sunweaver replied:
Hi Uli,

On Sun Jun 19 22:34:01 2016 GMT+0200, Ulrich Sibiller wrote:

> In nx-X11/programs/Xserver/hw/nxagent/NXextension.c:
> 
> > @@ -296,7 +296,8 @@ CloseDownExtensions()
> > 
> > ```
> >  for (i = NumExtensions - 1; i >= 0; i--)
> >  {
> > ```
> > -   (\* extensions[i]->CloseDown)(extensions[i]);
> > -   if (extensions[i]->CloseDown)
> > -       (\* extensions[i]->CloseDown)(extensions[i]);
> >   Hmm, same code as in dix/extensions.c. Can't this be simplified somehow? 
> >   —

## This currently work in progress. See PR #120.

Sent from my Jolla

uli42 commented on nx-X11/programs/Xserver/randr/randr.c:
```diff
@@ -234,8 +301,15 @@ RRInit(void)
+    if (!dixRegisterPrivateKey(&rrPrivKeyRec, PRIVATE_SCREEN, 0))
+        return FALSE;
+#endif                          /* !defined(NXAGENT_SERVER) */
+
```
This came here through https://cgit.freedesktop.org/xorg/xserver/commit/randr/randr.c?id=faeebead7bfcc78535757ca7acc1faf7554c03b7  Are you sure we do not need it for nxagent?

sunweaver replied:
This is covered a bit further up:

```
Bool
RRInit(void)
{
    if (RRGeneration != serverGeneration) {
#ifdef NXAGENT_SERVER
        if ((rrPrivIndex = AllocateScreenPrivateIndex()) < 0)
            return FALSE;
#endif
        if (!RRModeInit())
            return FALSE;
        if (!RRCrtcInit())
            return FALSE;
        if (!RROutputInit())
            return FALSE;
        if (!RRProviderInit())
            return FALSE;
        RRGeneration = serverGeneration;
    }
#ifndef NXAGENT_SERVER
    if (!dixRegisterPrivateKey(&rrPrivKeyRec, PRIVATE_SCREEN, 0))
        return FALSE;
#endif                          /* !defined(NXAGENT_SERVER) */

    return TRUE;
}
```

... by the call to AllocateScreenPrivateIndex().

uli42 commented on nx-X11/programs/Xserver/randr/randr.c:
```diff
@@ -309,7 +387,13 @@ RRFreeClient(void *data, XID id)
+#else                           /* !defined(NXAGENT_SERVER) */
     pHead = (RREventPtr *) LookupIDByType(pWin->drawable.id, RREventType);
+#endif                          /* !defined(NXAGENT_SERVER) */
+
```
Those lookups are called deprecated here: https://cgit.freedesktop.org/xorg/xserver/commit/randr/randr.c?id=faeebead7bfcc78535757ca7acc1faf7554c03b7

So _maybe_ we can switch those right now to the new way?

sunweaver replied:
The patch you are referencing is part of the privates ABI/API change in X.org. The change is massive. For this RandR backport, I'd rather stick to the old approach handling privates in nx-libs.

However, I have also been playing with the thought of backporting the privates ABI/API, but this is a bunch of work and also includes introducing the Xace extension into nx-libs.

(Something we would gain while rebasing, anyway...)

uli42 commented on nx-X11/programs/Xserver/randr/randr.c:
```diff
@@ -423,36 +572,147 @@ TellChanged(WindowPtr pWin, void *value)
+
+    mastersp->changed = TRUE;
+}
+
```
Could be simplified and replaced by two lines:

```
rrScrPriv(pScreen);
pScrPriv->changed = TRUE;
```

So one #ifndef would be enough.

sunweaver replied:
Ok, changed locally.

uli42 commented on nx-X11/programs/Xserver/randr/randrstr.h:
```diff
@@ -317,43 +425,132 @@ extern DevPrivateKey RRClientPrivateKey;
+
+#define VERIFY_RR_PROVIDER(id, ptr, a)\
+    {\
+	ptr = LookupProvider (client, id, a);\
```
We have dixLookupResourceByType in the 3.6 code base. Why not use it?

uli42 commented on nx-X11/programs/Xserver/randr/rrcrtc.c:
```diff
@@ -174,16 +198,24 @@ RRCrtcNotify(RRCrtcPtr crtc,
             if (crtc->numOutputs)
+#ifndef NXAGENT_SERVER
+                newoutputs = reallocarray(crtc->outputs,
+                                          numOutputs, sizeof(RROutputPtr));
```
Why not provide helper functions for (re)allocarray for now? This would save a couple of ifndefs...

uli42 commented on nx-X11/programs/Xserver/randr/rrcrtc.c:
```diff
@@ -225,30 +269,24 @@ RRDeliverCrtcEvent(ClientPtr client, WindowPtr pWin, RRCrtcPtr crtc)
+        .type = RRNotify + RREventBase,
+        .subCode = RRNotify_CrtcChange,
+#ifdef NXAGENT_SERVER
+        .sequenceNumber = client->sequence,
```
Looking around in the upstream code shows that sequence numbers are only set for WriteToClient() calls but not for WriteEventsToClient() calls. So I suspect that this is not needed here. Unsure if it will do any harm...

There are a few more places where this code is inserted.

Update: this patch explains that the sequence stuff is done automatically in WriteEventsToClient(): https://cgit.freedesktop.org/xorg/xserver/commit/randr/rrproperty.c?id=8033fb6c9792820a82fbdff6a14ff8a7a141ba74

As our current version does not do this we need to set these sequence numbers for now...

sunweaver replied:
Ok. Only saw the update for this comment now. Glad we agree here.

It actually caused me two days of debugging to find out that old ABI versions of WriteEventsToClient() also required the sequenceNumber to be set by the Xserver code.

If the squenceNumber is not being set as done above, the X11 clients crash whenever receiving an event from RandR in nxagent (been there, done that).

uli42 commented on nx-X11/programs/Xserver/randr/rrcrtc.c:
```diff
@@ -269,6 +307,302 @@ RRCrtcPendingProperties(RRCrtcPtr crtc)
+
+/* overlapping counts as adjacent */
+static Bool
+crtcs_adjacent(const RRCrtcPtr a, const RRCrtcPtr b)
```
great, could be helpful for the xinerama code

uli42 commented on nx-X11/programs/Xserver/randr/rrdispatch.c:
```diff
@@ -21,48 +21,62 @@
-
-#define SERVER_RANDR_MAJOR	1
-#define SERVER_RANDR_MINOR	2
+#ifndef NXAGENT_SERVER
```
I think it would be better to not abuse NXAGENT_SERVER for stuff that is missing in the used Xorg base. Maybe these places should be changed to XORG_TOO_OLD or something. This way we can identify them easier as the Xorg code gets updated.

sunweaver replied:
This is true. I had discussed that with @Ionic earlier while he backported the Composite extension code (0.2 -> 0.4) to nx-libs.

His reasoning was that NoMachine used NXAGENT_SERVER for indicating missing Xserver features.

I agree that we should have some XORG_TOO_OLD macro or such, as well. I won't do this for this pull request but may provide a patch for that later. See #144.

uli42 commented on nx-X11/programs/Xserver/randr/rrmode.c:
```diff
@@ -113,15 +117,11 @@ RRModeCreate(xRRModeInfo * modeInfo, const char *name, ScreenPtr userScreen)
-    fprintf(stderr,
-            "RRModeCreate: num_modes [%d] new mode [%s] ([%p]) refcnt [%d]\n",
-            num_modes, mode->name, mode, mode->refcnt);
-#endif
```
this ismissing int the new code

sunweaver replied:
also re-added in rrmode.c

uli42 commented on nx-X11/programs/Xserver/randr/rrmode.c:
```diff
@@ -152,26 +152,11 @@ RRModeGet(xRRModeInfo * modeInfo, const char *name)
-            fprintf(stderr,
-                    "RRModeGet: return existing mode [%s] ([%p]) refcnt [%d]\n",
-                    mode->name, mode, mode->refcnt);
-#endif
```
also missing

sunweaver replied:
DEBUG output re-added to rrmode.c (line 155ff).

uli42 commented on nx-X11/programs/Xserver/randr/rrmode.c:
```diff
@@ -152,26 +152,11 @@ RRModeGet(xRRModeInfo * modeInfo, const char *name)
-    }
-#else
     return RRModeCreate(modeInfo, name, NULL);
-#endif
```
also missing. I needed them to debug refcnt issues for xinerama

sunweaver replied:
DEBUG output also re-added in rrmode.c (lines apprx. 164 - 174).

uli42 commented:
On Sat, Jun 18, 2016 at 7:52 PM, Mike Gabriel notifications@github.com wrote:

> @uli42: Please consider reviewing this PR, as well. It is big and we shall consider your review work on this as paid time. If you feel, that this is too much work or not doable for you, please let me know.

OK, I have looked through it and commented some stuff. I think at
least the (re)alloc calls should be implemented. This would make the
code much cleaner (and the diff much smaller).

The 'monitor' stuff looks pretty helpful for the xinerama code. I
think we can improve (and simplify) the xinerama code with that.
However, I am not sure if I can come up with code for that soon.

sunweaver commented:
On  Mo 20 Jun 2016 00:40:02 CEST, Ulrich Sibiller wrote:

> On Sat, Jun 18, 2016 at 7:52 PM, Mike Gabriel  
> notifications@github.com wrote:
> 
> > @uli42: Please consider reviewing this PR, as well. It is big and  
> > we shall consider your review work on this as paid time. If you  
> > feel, that this is too much work or not doable for you, please let  
> > me know.
> 
> OK, I have looked through it and commented some stuff. I think at

Thanks for doing a review so quickly.

> least the (re)alloc calls should be implemented. This would make the
> code much cleaner (and the diff much smaller).

I actually have a branch here locally that backports various ABI  
upgrades of latest X.org to nx-libs. This basically comes from a  
lessons-learned experience when backporting RandR 1.5. I plan to  
backport more extensions, actually, for the 3.6.x series. Maybe not  
everything for 3.6.0, but definitely for later point releases of 3.6.x.

 From experience with earlier PRs, I want to suggest this for this PR:

   o let's get this PR merged rather sooner than later (because it is  
a big one)
   o let's fix up the (re)alloc thing globally for all occurrences in
     nx-libs later (but before 3.6.0 release)
   o let's fix more ABI issues (xfree -> free, pScreen->root, etc.), as well.

> The 'monitor' stuff looks pretty helpful for the xinerama code. I
> think we can improve (and simplify) the xinerama code with that.
> However, I am not sure if I can come up with code for that soon.

Very good.

I will now go through all your comments and fix things locally.

Also, I will split up changes in Xserver/randr/ from changes in the  
rest of the Xserver code. I should have done that in the first place,  
actually.

BTW: Also, the new version of Xserver/randr in nx-libs gives a nice  
"diff -w" when compared to X.org's Xserver/randr code.

Mike

## 

DAS-NETZWERKTEAM
mike gabriel, herweg 7, 24357 fleckeby
mobile: +49 (1520) 1976 148
landline: +49 (4354) 8390 139

GnuPG Fingerprint: 9BFB AEE8 6C0A A5FF BF22  0782 9AF4 6B30 2577 1B31
mail: mike.gabriel@das-netzwerkteam.de, http://das-netzwerkteam.de

uli42 commented:
ok. 

BTW: I use xxdiff most of the time

sunweaver commented:
@uli42: ok, I have reworked this whole PR.

The RANDR backport has not change except for those changes proposed by you.

I have not yet done the reallocarray change (because this PR is too massive already and that change is not RANDR specific).

But I have split up all necessary Xserver changes into individual commits and references the X.org origins behind them.

Furthermore, I fully removed the WindowTable array, because introducing pScreen->root allows for that now. See https://github.com/ArcticaProject/nx-libs/pull/137/commits/12f16d353cad9cbbe9092a7d6a08cf0a1f473f17.

Please take a last look. If ok with those changes, I will rebase the top most RANDR patches (   4302b63 and     82614e5) and merge them into 05a133f.

Furthermore, I just saw (while writing this) that the render patches required for the RANDR backport come after the actual backport. That order I will change after approval, too.

Thanks for taking a second look.

uli42 commented on nx-X11/programs/Xserver/hw/nxagent/NXevents.c:
```diff
@@ -2307,7 +2307,7 @@ WindowsRestructured()
 #ifdef PANORAMIX
 /* This was added to support reconfiguration under Xdmx.  The problem is
- * that if the 0th screen (i.e., WindowTable[0]) is moved to an origin
+ * that if the 0th screen (i.e., screenInfo.screens[0]->root) is moved to an origin
```
looks like line is too wide here

sunweaver replied:
> looks like line is too wide here

Fixed locally.

uli42 commented on nx-X11/programs/Xserver/hw/nxagent/Windows.h:
```diff
@@ -167,7 +167,7 @@ extern int nxagentWindowPrivateIndex;
-  nxagentRefreshWindows(WindowTable[0]);\
+  nxagentRefreshWindows(screenInfo.screens[0]->root);\
 } while (0)
 
```
Hmm, what's that good for? Looks strange..

uli42 commented on nx-X11/programs/Xserver/include/scrnintstr.h:
```diff
@@ -549,6 +549,7 @@ typedef struct _Screen {
     short       	numVisuals;
     VisualPtr		visuals;
+    WindowPtr		root;
     int			WindowPrivateLen;
```
Formatting broken

sunweaver replied:
Formatting in scrnintstr.h (WindowPtr root;) is not really broken. The breakage is caused due to Git displaying tabs in a non-consistent way in diffs.

uli42 commented:
Looks good to me. I have added some comments regarding formatting, but feel free to ignore them.

uli42 commented:
Hmm, I have tried to compile the current state and see this:

Handlers.c: In function ‘nxagentShadowBlockHandler’:
Handlers.c:755:0: error: unterminated #ifdef
 #ifdef TEST
 ^
Handlers.c:754:3: error: expected declaration or statement at end of input
   }
   ^
Handlers.c:736:15: warning: unused variable ‘height_’ [-Wunused-variable]
   int width_, height_;
               ^
Handlers.c:736:7: warning: unused variable ‘width_’ [-Wunused-variable]
   int width_, height_;
       ^
Handlers.c:735:7: warning: unused variable ‘suspended’ [-Wunused-variable]
   int suspended = 0;
       ^
Handlers.c:734:7: warning: unused variable ‘changed’ [-Wunused-variable]
   int changed;
       ^
Handlers.c:732:25: warning: unused variable ‘zero’ [-Wunused-variable]
   static struct timeval zero;
                         ^
make[7]: **\* [Handlers.o] Fehler 1
make[7]: Verzeichnis
»/home/uli/work/nx/ArcticaProject/nx-libs-sunweaver/nx-X11/programs/Xserver/hw/nxagent«
wird verlassen

Uli

uli42 commented:
Fix:

diff --git a/nx-X11/programs/Xserver/hw/nxagent/Handlers.c
b/nx-X11/programs/Xserver/hw/nxagent/Handlers.c
index ba40342..62cc549 100644
--- a/nx-X11/programs/Xserver/hw/nxagent/Handlers.c
+++ b/nx-X11/programs/Xserver/hw/nxagent/Handlers.c
@@ -756,6 +756,7 @@ void nxagentShadowBlockHandler(void \* data, struct
timeval **timeout, void \* mas
   else if (0 == nxagentOption(SleepTime)) {
     fprintf(stderr, "nxagentBlockHandler: not sleeping for slowdown.\n");
   }
+#endif

   #ifndef **CYGWIN32**
